glean-core/tests/dual_labeled_counter.rs

```diff
@@ -0,0 +1,527 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+mod common;
+use crate::common::*;
+
+use serde_json::json;
+
+use glean_core::metrics::*;
+use glean_core::storage::StorageManager;
+use glean_core::{test_get_num_recorded_errors, ErrorType};
+use glean_core::{CommonMetricData, Lifetime};
+
+#[test]
+fn can_create_dual_labeled_counter_metric() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        Some(vec!["key1".into()]),
+        Some(vec!["category1".into()]),
+    );
+
+    let metric = dual_labeled_counter.get("key1", "category1");
+    metric.add_sync(&glean, 1);
+
+    let snapshot = StorageManager
+        .snapshot_as_json(glean.storage(), "store1", true)
+        .unwrap();
+
+    assert_eq!(
+        json!({
+            "dual_labeled_counter": {
+                "telemetry.dual_labeled_counter": {
+                    "key1": {
+                        "category1": 1
+                    }
+                }
+            }
+        }),
+        snapshot
+    );
+}
+
+#[test]
+fn can_use_multiple_labels() {
+    let (glean, _t) = new_glean(None);
+    let dual_labeled_counter = DualLabeledCounterMetric::new(
+        CommonMetricData {
+            name: "dual_labeled_counter".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::Ping,
+            ..Default::default()
+        },
+        None,
+        None,
+    );
+
+    let metric = dual_labeled_counter.get("key1", "category1");
+    metric.add_sync(&glean, 1);
+
+    let metric = dual_labeled_counter.get("key1", "category2");
+    metric.add_sync(&glean, 2);
+
+    let metric = dual_labeled_counter.get("key2", "category1");
+    metric.add_sync(&glean, 3);
+
+    let metric = dual_labeled_counter.get("key2", "category2");
```

Comment:
To cover one more case you could make this `category3` to show that the categories are independent per key.